pkg/minio/minio.go

```diff
@@ -0,0 +1,418 @@
+			endpoint, err := m.getEndpoint(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get Minio endpoint: %v", err)
+			}
+
+			if err := checkServiceConnectivity(endpoint); err == nil {
+				break
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timeout waiting for Minio service to be ready")
+		case <-time.After(waitRetry):
+			// Retry after some seconds
+		}
+	}
+
+	return nil
+}
+
+func checkServiceConnectivity(serviceEndpoint string) error {
+	conn, err := net.DialTimeout("tcp", serviceEndpoint, 2*time.Second)
+	if err != nil {
+		return fmt.Errorf("failed to connect to %s: %v", serviceEndpoint, err)
+	}
+	defer conn.Close()
+	return nil // success
+}
+
+func (m *Minio) createPVC(ctx context.Context, pvcName string, storageSize string) error {
+	storageQt, err := resource.ParseQuantity(storageSize)
+	if err != nil {
+		return fmt.Errorf("failed to parse storage size: %v", err)
+	}
+
+	pvcClient := m.Clientset.CoreV1().PersistentVolumeClaims(m.Namespace)
+
+	// Check if PVC already exists
+	_, err = pvcClient.Get(ctx, pvcName, metav1.GetOptions{})
+	if err == nil {
+		logrus.Debugf("PersistentVolumeClaim `%s` already exists.", pvcName)
+		return nil
+	}
+
+	// Create a simple PersistentVolume if no suitable one is found
+	pvList, err := m.Clientset.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list PersistentVolumes: %v", err)
+	}
+
+	var existingPV *v1.PersistentVolume
+	for _, pv := range pvList.Items {
+		// Not sure if this condition is ok
+		if pv.Spec.Capacity[v1.ResourceStorage].Equal(storageQt) {
+			existingPV = &pv
+			break
+		}
+	}
+
+	if existingPV == nil {
+		// Create a simple PV if no existing PV is suitable
+		_, err = m.Clientset.CoreV1().PersistentVolumes().Create(ctx, &v1.PersistentVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "minio-pv-",
+			},
+			Spec: v1.PersistentVolumeSpec{
+				Capacity: v1.ResourceList{
+					v1.ResourceStorage: storageQt,
+				},
+				AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+				PersistentVolumeSource: v1.PersistentVolumeSource{
+					HostPath: &v1.HostPathVolumeSource{
+						Path: "/tmp/minio-pv", // Replace with your desired host path
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create PersistentVolume: %v", err)
+		}
+
+		logrus.Debugf("PersistentVolume `%s` created successfully.", existingPV.Name)
+	}
+
+	// Create PVC with the existing or newly created PV
+	pvc := &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pvcName,
+			Namespace: m.Namespace,
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+			Resources: v1.VolumeResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: storageQt,
+				},
+			},
+		},
+	}
+
+	_, err = pvcClient.Create(ctx, pvc, metav1.CreateOptions{})
```

Comment:
why are we not allowing options in signature func?

Reply:
Ideally, we need to refactor this in the future to accomodate such use-cases

Reply:
done